source/change-streams.rst

```diff
@@ -41,7 +41,62 @@ Terms
 Resumable Error
 ^^^^^^^^^^^^^^^
 
-Any error encountered which is not a server error, with the exception of server responses with the message “not master” or error code 43 (cursor not found).  An example might be a timeout error, or network error.
+An error is considered resumable if it meets any of the following criteria:
+
+- any error encountered which is not a server error (e.g. a timeout error or
+  network error)
+
+- a server error response with any of the following error codes:
+
+  .. list-table::
+    :header-rows: 1
+
+    * - Error Name
+      - Error Code
+    * - HostUnreachable
+      - 6
+    * - HostNotFound
+      - 7
+    * - NetworkTimeout
+      - 89
+    * - SocketException
+      - 9001
+    * - NotMaster
+      - 10107
+    * - NotMasterNoSlaveOk
+      - 13435
+    * - NotMasterOrSecondary
+      - 13436
+    * - InterruptedDueToReplStateChange
+      - 11602
+    * - PrimarySteppedDown
+      - 189
+    * - ExceededTimeLimit
+      - 50
+    * - RetryChangeStream
+      - 234
+
+- a server error response without an error code or one different from those
+  listed above, but with an error message containing the substring "not
+  master" or "node is recovering"
+
+- *any* server error response from a getMore command excluding those
+  containing the following error codes
+
+  .. list-table::
+    :header-rows: 1
+
+    * - Error Name
+      - Error Code
+    * - Interrupted
```

Comment:
We added a new one in 4.0: https://github.com/mongodb/mongo/blob/0c1402ee3f10eaf9d6d989d814bc3113a380091e/src/mongo/base/error_codes.err#L237

Reply:
Also, 'CursorNotFound' and CappedPositionLost are hopeless to retry.

Reply:
CursorNotFound is the O.G. retryable error actually - we retry the change stream with a resume token and get a new cursor.

Reply:
> Also, 'CursorNotFound' and CappedPositionLost are hopeless to retry.

If they are indeed hopeless, the retry attempt to reestablish the change stream cursor should fail with "resume token is too old".

Reply:
Shouldn't CursorNotFound be resumable? We can't use that cursor, but resuming creates a new cursor.

Reply:
My mistake - I was thinking of retrying the literal getMore command. I now remember we're talking about retrying/resuming the change stream. Forget about CursorNotFound - I think the others would still be good to include though. CappedPositionLost implies you fell off the back of the oplog, and CursorKilled implies some other actor killed your cursor - similar to Interrupted.